File: src/extensions/reader.ts
import { execSync } from "node:child_process";
import { existsSync, readdirSync } from "node:fs";
import type { ResolvedEditor } from "../editors/types";

export interface ExtensionInfo {
  id: string;
  source: "cli" | "dir-scan";
}

export interface EditorExtensions {
  editor: ResolvedEditor;
  extensions: ExtensionInfo[];
}

function readViaCli(cli: string): ExtensionInfo[] | null {
  try {
    const output = execSync(`${cli} --list-extensions`, {
      encoding: "utf-8",
      timeout: 15000,
      stdio: ["ignore", "pipe", "ignore"],
    });
    return output
      .split("\n")
      .map((l) => l.trim())
      .filter(Boolean)
      .map((id) => ({ id: id.toLowerCase(), source: "cli" as const }));
  } catch {
    return null;
  }
}

function readViaDirScan(extensionsPath: string): ExtensionInfo[] | null {
  if (!existsSync(extensionsPath)) return null;

  try {
    const entries = readdirSync(extensionsPath, { withFileTypes: true });
    return entries
      .filter((e) => e.isDirectory())
      .map((e) => e.name)
      .filter((name) => !name.startsWith("."))
      .map((name) => {
        const match = name.match(/^(.+?)-\d+\.\d+\.\d+/);
        const id = match ? match[1] : name;
        return { id: id.toLowerCase(), source: "dir-scan" as const };
      })
      .filter(
        (ext, i, arr) => arr.findIndex((e) => e.id === ext.id) === i
      );
  } catch {
    return null;
  }
}

export function readExtensions(editor: ResolvedEditor): EditorExtensions {
  if (editor.cliAvailable) {
    const cliResult = readViaCli(editor.cli);
    if (cliResult && cliResult.length > 0) {
      return { editor, extensions: cliResult };
    }
  }

  const dirResult = readViaDirScan(editor.extensionsPath);
  if (dirResult) {
    return { editor, extensions: dirResult };
  }

  return { editor, extensions: [] };
}

export function readAllExtensions(
  editors: ResolvedEditor[]
): EditorExtensions[] {
  return editors.map(readExtensions);
}

File: src-tauri/src/extensions/reader.rs
use std::collections::HashSet;
use std::process::Command;

use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::editors::ResolvedEditor;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtensionInfo {
    pub id: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EditorExtensions {
    pub editor: ResolvedEditor,
    pub extensions: Vec<ExtensionInfo>,
}

fn read_via_cli(editor: &ResolvedEditor) -> Option<Vec<ExtensionInfo>> {
    let output = Command::new(&editor.cli)
        .arg("--list-extensions")
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout);

    let result = stdout
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(|line| ExtensionInfo {
            id: line.to_ascii_lowercase(),
            source: "cli".to_string(),
        })
        .collect::<Vec<_>>();

    Some(result)
}

fn read_via_dir_scan(editor: &ResolvedEditor) -> Option<Vec<ExtensionInfo>> {
    let path = std::path::Path::new(&editor.extensions_path);
    if !path.exists() {
        return None;
    }

    let regex = Regex::new(r"^(.+?)-\d+\.\d+\.\d+").ok()?;
    let entries = std::fs::read_dir(path).ok()?;

    let mut dedup = HashSet::new();
    let mut result = Vec::new();

    for entry in entries.flatten() {
        let file_type = match entry.file_type() {
            Ok(value) => value,
            Err(_) => continue,
        };

        if !file_type.is_dir() {
            continue;
        }

        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') {
            continue;
        }

        let id = regex
            .captures(&name)
            .and_then(|capture| capture.get(1).map(|matched| matched.as_str().to_string()))
            .unwrap_or(name)
            .to_ascii_lowercase();

        if dedup.insert(id.clone()) {
            result.push(ExtensionInfo {
                id,
                source: "dir-scan".to_string(),
            });
        }
    }

    Some(result)
}

pub fn read_extensions(editor: &ResolvedEditor) -> EditorExtensions {
    if editor.cli_available {
        if let Some(cli_result) = read_via_cli(editor) {
            if !cli_result.is_empty() {
                return EditorExtensions {
                    editor: editor.clone(),
                    extensions: cli_result,
                };
            }
        }
    }

    if let Some(dir_result) = read_via_dir_scan(editor) {
        return EditorExtensions {
            editor: editor.clone(),
            extensions: dir_result,
        };
    }

    EditorExtensions {
        editor: editor.clone(),
        extensions: Vec::new(),
    }
}

pub fn read_all_extensions(editors: &[ResolvedEditor]) -> Vec<EditorExtensions> {
    editors.iter().map(read_extensions).collect()
}
